Allow connecting/disconnecting from networks in UI, closes #631

MaxLeiter opened the pull request:
Few thoughts that need to be worked out:

1. Not sure how I feel about icons
<table>
  <tr>
    <td>
<img width="247" alt="screen shot 2018-03-17 at 2 06 24 pm" src="https://user-images.githubusercontent.com/8675906/37559961-7e4cba02-29ec-11e8-9f9c-ccc945d064db.png">
</td>
    <td>
<img width="229" alt="screen shot 2018-03-17 at 2 06 29 pm" src="https://user-images.githubusercontent.com/8675906/37559958-77793354-29ec-11e8-91cc-bcab100b9f39.png">

</td>
  </tr>
</table>

2. Should status be saved server-side? Add a `connected` field to `Network`?
3. More I'm sure I forgot
4. Need to deal with add-channel UI when disconnected. Thoughts?

christer88 commented:
I guess this PR will close https://github.com/thelounge/thelounge/issues/1112 because we now have "Disconnect" to only disconnect, and "Remove" to remove network from json config file? If so, thanks @MaxLeiter!

McInkay commented:
My poor #1878

creesch commented:
> Need to deal with add-channel UI when disconnected. Thoughts?

Leave channels with logging in place, disable input field for channels on disconnected networks.

xPaw commented:
There's no need to disable the input because lounge will join it after connecting.

MaxLeiter commented:
Ready to go?

McInkay commented:
Oh please... Mine first.

MaxLeiter commented:
Fine with me, will be a simple rebase

On Mon, Feb 26, 2018 at 12:22 PM Al McKinlay <notifications@github.com>
wrote:

> Oh please... Mine first.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/thelounge/thelounge/pull/2114#issuecomment-368637211>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIRiQmhRZMc9x9SGyFfGjOM7gQT4G6xsks5tYxKHgaJpZM4SSHOW>
> .
>

McInkay commented:
Well let's see what @astorije says about mine after my refactor earlier today. It's just rebasing mine on top of changes does ideas the likelihood of regressions. I'm not going to ask you to wait weeks.

Haarolean commented:
Would be nice to be able to change network settings via UI as well. 
https://github.com/thelounge/thelounge/issues/179 is related.

McInkay commented:
@Haarolean This is unrelated to the PR, please keep relevant discussion to their own tickets, as you have commented on the other issue. Thanks.

astorije commented:
I don't think "Disconnect" is intentional here, right?

<img width="292" alt="screen shot 2018-03-04 at 20 35 28" src="https://user-images.githubusercontent.com/113730/36953832-a17ea1f8-1feb-11e8-8926-a1d2f40ab97c.png">

astorije commented:
Do we need both entries "Disconnect" and "Remove" (which also disconnects)?

MaxLeiter commented:
@astorije: Remove deletes, disconnect disconnects

astorije commented:
Yes I understand that. Remove also disconnects (right?). But, do we actually need that?

I sort of understand the need for `Connect` (actually, `Reconnect`?) because connection with server can be lost from time to time, and there are cases when we don't automatically reconnect (though we should limit this as much as we can), but why "Disconnect" on top of "Remove"?
It seems to me that the main purpose of TL is to stay connected at all time, reconnect whenever possible, etc. so I'm just trying to understand the need behind this, in case we're trying to solve the wrong problem.

MaxLeiter commented:
There are times where you want to disconnect so you can reconnect, and there are times when you want to delete the network. Sometimes you want to save the network info but temporally leave it. Pretty common in clients, afaik

astorije commented:
What about a "Reconnect" instead that disconnects + reconnects? That seems to me to be the end-goal here.

McInkay commented:
> What about a "Reconnect" instead that disconnects + reconnects? That seems to me to be the end-goal here.

It's the end goal for some of it, sure, but there is definitely a use case for an irc client that you maybe don't always want to be connected to, and making "Disconnect" and "Connect" instead of "Reconnect" allows both use cases.

christer88 commented:
I agree with @McInkay!

MaxLeiter commented:
Updated links to FA 5, rebased to master, updated screenshots

McInkay commented:
This will need to be refactored now that #1878 is merged, and use the new structure for adding context menu items. If you have any questions about how this works, or anything is unclear about what I've changed then give me a shout @MaxLeiter

MaxLeiter commented:
@McInkay: working on updating to master.
2 questions
1) Where should I set `connected`? Currently, I set it in `check()`.
2) `displayName` does not seem to be updating. (As in, always shows "Connected"). Fixing number 1 properly may fix this inherently (like if I make two different `ContextMenuItem`'s)


``` javascript
	let connected;
	contextMenuFactory.addContextMenuItem({
		check: (target) => {
			// not ideal to have the line below here. Not sure where else to place it.
			connected = !target.closest("section").hasClass("not-connected");
			console.log(target.closest("section").hasClass("not-connected"));
			return target.hasClass("lobby");
		},
		className: connected ? "disconnect" : "connect",
		displayName: connected ? "Disconnect" : "Connect",
		data: (target) => target.data("id"),
		callback: (itemData) => {
			if (!connected) {
				socket.emit("input", {
					target: itemData,
					text: "/connect",
				});
			} else {
				socket.emit("input", {
					target: itemData,
					text: "/disconnect",
				});
			}
		},
	});
```

McInkay commented:
@MaxLeiter 

> Where should I set connected? Currently, I set it in check()

I think you'll need to set it onto the object, otherwise the scope won't work? Maybe I'm wrong.

> displayName does not seem to be updating.

Yeah, this is because you are passing through the string, not a function. Change it to this:

```
className: () => connected ? "disconnect" : "connect",
displayName: () => connected ? "Disconnect" : "Connect",
```

This way the class and display name are recalculated everytime it creates a menu item, rather than just when you create it.

Thanks for the advice, I'll make sure to put this more clearly into the docs when I write them.

Hope that makes sense.

MaxLeiter commented:
Rebased + updated, running into a slight issue with `className` though -- should not need arrow notation for it to function

McInkay commented:
> running into a slight issue with className though -- should not need arrow notation for it to function

Yes, you absolutely should.

If you are passing through `connected ? "Disconnect" : "Connect"` to the function, that is processed by Javascript and sent as a String. Therefore it'll always be the same. You need to send a function for it to ever change.

astorije commented on client/js/lounge.js:
```diff
@@ -330,14 +330,48 @@ $(function() {
+			clickedNetwork = target;
+			return target.hasClass("lobby");
+		},
+		className: () => isConnected(clickedNetwork) ? "connect" : "disconnect", // eslint-disable-line no-confusing-arrow
```
I'm not fan of disabling stuff like this. Use the following instead:

```js
className() {
  return isConnected(clickedNetwork) ? "connect" : "disconnect";
},
```

MaxLeiter commented:
Yeah just haven’t got to it yet

MaxLeiter commented:
Should be good for reviews

astorije commented:
Thanks @MaxLeiter!

astorije commented on client/js/lounge.js:
```diff
@@ -330,14 +330,50 @@ $(function() {
+	};
+
+	let clickedNetwork;
+	contextMenuFactory.addContextMenuItem({
```
@xPaw raised a good point, could you split this into 2 items? I believe you'd have your `isConnected(target)` in `check`. That would simplify things like `className` 😅

xPaw replied:
Also move all this code into context menu factory.

MaxLeiter replied:
That complicates things like `isConnected()`, will need to play with this

MaxLeiter commented:
Okay, split up connect/disconnect and moved to factory.

Also moved other context code left-over from lounge.js -> factory, but that required moving `closeChan` to `utils.js` so the Factory could use it. I don't care if we leave the close chan context item in `lounge.js` or not, just let me know.
